Support remote write v2 by converting request

SungJin1212 opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
-->

This PR supports Prometheus remote write 2.0 by converting the v2 request to v1 at the API. 

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

yeya24 commented:
Looks promising. Thanks!

FYI we have https://github.com/prometheus/client_golang/pull/1658 which exports remote write handler. Not a blocker for this PR but we should keep it on our radar to switch to use the library

SungJin1212 commented:
@yeya24 
Thanks for letting me know.
Should we make the issue to track it?

alanprot commented:
Maybe we can open a issue for someone give a try to use the client_golang handler even before it get merged so we can give feedback on the open PR. Changing that handler after is merged probably will be more difficult as it could potentially break all projects that are already using it.

SungJin1212 commented:
@alanprot 
I added a comment here: https://github.com/cortexproject/cortex/issues/6324

yeya24 commented:
I took a breif look at https://github.com/prometheus/client_golang/pull/1658. Left some comments there and we have some changes Cortex specific that might not make sense for Prometheus. I think we are ok to proceed with this PR first.

SungJin1212 commented:
@yeya24 
Thanks. I read it, and it would be good if we could reuse its functions!

yeya24 commented on pkg/util/push/push.go:
```diff
@@ -28,31 +49,232 @@ func Handler(maxRecvMsgSize int, sourceIPs *middleware.SourceIPExtractor, push F
+			}
+		case config.RemoteWriteProtoMsgV2:
+			var req writev2.Request
+			err := util.ParseProtoReader(ctx, r.Body, int(r.ContentLength), maxRecvMsgSize, &req, util.RawSnappy)
```
@alanprot @danielblando
I wonder if we want to introduce a feature flag to control the behavior for RW v2 request. We can either ignore the request or convert to v1 and in the future maybe just accept as is.

yeya24 replied:
Let's add a feature flag for the purpose of rollout. If RW 2.0 conversion is enabled right away, then Ingesters need to be rolled out first because of the protocol change to return stats. If we want to rollout Ingester and Distributor the same time then things can go wrong without a feature flag.

yeya24 commented on pkg/ingester/ingester.go:
```diff
@@ -1401,7 +1401,13 @@ func (i *Ingester) Push(ctx context.Context, req *cortexpb.WriteRequest) (*corte
-	return &cortexpb.WriteResponse{}, nil
+	writeResponse := &cortexpb.WriteResponse{
+		Samples:    int64(succeededSamplesCount),
+		Histograms: int64(succeededSamplesCount), //TODO(Sungjin1212): Should we count histogram?
```
How is this implemented in Prometheus. Why we don't count histogram here?

SungJin1212 replied:
I left `TODO` since we are counting histograms just as we count the sample.
But, the Prometheus is counting native histogram https://github.com/prometheus/prometheus/blob/main/storage/remote/write_handler.go#L424.

SungJin1212 replied:
How about starting to count `histogram` when we introduce `PushV2`?

yeya24 replied:
I don't understand the concern here. There is nothing prevent us doing it. We should count native histograms

SungJin1212 replied:
yes, I agree with counting native histograms by changing to `Histograms: int64(nativeHistogramCount)`.
My concern is we are counting samples instead of `native histograms`
https://github.com/cortexproject/cortex/blob/master/pkg/ingester/ingester.go#L1269

yeya24 replied:
`succeededSamplesCount` is just an internal state/metric we track. There is nothing user facing. We can either change it or introduce new variables to count normal samples and native histogram samples.

SungJin1212 replied:
Can I make PR to introduce `succeededHistogramCount` and `failedHistogramCount` first?

yeya24 replied:
I am fine to split but why it needs a separate PR? We can just add a new int64 variable to count succeeded histograms

SungJin1212 replied:
Maybe some changes are needed like we are tracking ingestionRate by calculating `succeededSamplesCount + ingestedMetadata`.
Also, we can introduce new metrics like `cortex_ingester_ingested_native_histograms_total` and `cortex_ingester_ingested_histograms_failures_total`.
WDYT?

yeya24 replied:
OK. I am ok to add the new metrics. They are not blocking this PR so can be done either now or after this change.

yeya24 commented on pkg/distributor/distributor.go:
```diff
@@ -816,7 +824,7 @@ func (d *Distributor) doBatch(ctx context.Context, req *cortexpb.WriteRequest, s
 		}
 
-		return d.send(localCtx, ingester, timeseries, metadata, req.Source)
+		return d.send(localCtx, ingester, timeseries, metadata, req.Source, stats)
```
Do we need to sum the samples pulled from `stats`? Now I see we just overwrite stats for every request.

SungJin1212 replied:
If so, isn't there a good chance the returned header value (`X-Prometheus-Remote-Write-Samples-Written`) would be multiple of samples in a write request?

yeya24 replied:
Yeah. With replication factor it is expected to have more samples. I think this is fine.

SungJin1212 commented:
@yeya24 
I added `-distributor.remote-write2-enabled` flags to configure whether the Distributor can accept PRW2.0.
I added the `TestIngesterRollingUpdate` e2e test, where the Distributor can accept PRW2.0 and the Ingester uses the v1.18.1 image.
The result is PRW2.0 push is a success, but the response header (`X-Prometheus-Remote-Write-xxx`) values are all "0".
Is it what you expect behavior?

yeya24 commented:
> The result is PRW2.0 push is a success, but the response header (X-Prometheus-Remote-Write-xxx) values are all "0".
Is it expecting behavior?

This doesn't sound like the right behavior. Is that what you got with this PR?

SungJin1212 commented:
Yes, the test condition is that Ingester uses a v1.18.1 image and Distributor uses a PRW2.0-implemented one. Then, the PRW2.0 push gets that result.
If Ingester and Distributor use the same images (PRW 2.0 implemented), then we can get the expected response header.

yeya24 commented:
>  Yes, the test condition is that Ingester uses a v1.18.1 image and Distributor uses a PRW2.0-implemented one. The PRW2.0 push then gets that result.

I see what you meant. Then it is expected to get that result if you use Ingester of old version and Distributor of new version.
That's why we introduce the PRW 2.0 feature flag in distributor to only enable PRW 2.0 request if backend Ingester is running the newer version.

SungJin1212 commented:
Should I add comments to `-distributor.remote-write2-enabled` so that the user can do a rolling update of the Ingesters first and then update the Distributor afterward?

yeya24 commented:
We can mention it in the flag description but I doubt if users really look at it.
I prefer to create a dedicated doc/guide for users to migrate to Prometheus 3.0

CharlieTLe commented:
Hello @SungJin1212, thank you for opening this PR.

There is a release in progress. As such, please rebase your CHANGELOG entry on top of the master branch and move the CHANGELOG entry to the top under `## master / unreleased`.

Thanks,
Charlie

yeya24 commented:
I think it is time to revisit this now : )

yeya24 commented:
Ping for review. @danielblando @alanprot @friedrichg

alanprot commented on pkg/util/push/push.go:
```diff
@@ -28,31 +51,240 @@ func Handler(maxRecvMsgSize int, sourceIPs *middleware.SourceIPExtractor, push F
+			contentType = appProtoContentType
+		}
+
+		msgType, err := parseProtoMsg(contentType)
 		if err != nil {
-			level.Error(logger).Log("err", err.Error())
-			http.Error(w, err.Error(), http.StatusBadRequest)
+			level.Error(logger).Log("Error decoding remote write request", "err", err)
+			http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
 			return
 		}
 
-		req.SkipLabelNameValidation = false
-		if req.Source == 0 {
-			req.Source = cortexpb.API
+		if msgType != config.RemoteWriteProtoMsgV1 && msgType != config.RemoteWriteProtoMsgV2 {
+			level.Error(logger).Log("Not accepted msg type", "msgType", msgType, "err", err)
+			http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
+			return
+		}
```
Thinking out loud:
Should we be in the safe side and fallback to v1 if the content-type is not expected?

I guess if right now u send a weird content type cortex will acccept? and after this change we would return "StatusUnsupportedMediaType"?

SungJin1212 replied:
The [PRW2.0](https://prometheus.io/docs/specs/prw/remote_write_spec_2_0/#unsupported-request-content) spec says the receiver should return `StatusUnsupportedMediaType`.

Would it be more nicer if we fallback to v1 when the `StatusUnsupportedMediaType ` case?

alanprot replied:
I think so? we don’t wanna to break v1 clients I guess ?  It’s more important I to know what v1 protocol says than v2 I think  … cause we cannot assume v2 protocol behavior if we don’t know what version the request is as the content type is unknown?

SungJin1212 replied:
I changed not to break existing behavior. Could you take a look?

alanprot commented on pkg/util/push/push.go:
```diff
@@ -28,31 +51,240 @@ func Handler(maxRecvMsgSize int, sourceIPs *middleware.SourceIPExtractor, push F
+			}
+		case config.RemoteWriteProtoMsgV2:
+			if remoteWrite2Enabled {
+				var req writev2.Request
```
Do you think we should also pool the "writev2.Request" like we do with the "PreallocWriteRequest", so we avoid lots of GC?

Maybe for this PR its ok to not do that but i imagine the GC will be quite high with v2 if we dont do that.

SungJin1212 replied:
yeah, we need to introduce pool like v1. Let's update it to #6324.

SungJin1212 replied:
I added benchmarks and here are the results:
```
Benchmark_Handler
Benchmark_Handler/PRW1_with_10_series
Benchmark_Handler/PRW1_with_10_series-10         	  109134	     10406 ns/op	   24970 B/op	     246 allocs/op
Benchmark_Handler/PRW2_with_10_series
Benchmark_Handler/PRW2_with_10_series-10         	   73774	     16280 ns/op	   33920 B/op	     374 allocs/op
Benchmark_Handler/PRW1_with_100_series
Benchmark_Handler/PRW1_with_100_series-10        	   13016	     90841 ns/op	  233753 B/op	    2319 allocs/op
Benchmark_Handler/PRW2_with_100_series
Benchmark_Handler/PRW2_with_100_series-10        	    9264	    131270 ns/op	  297826 B/op	    3350 allocs/op
Benchmark_Handler/PRW1_with_500_series
Benchmark_Handler/PRW1_with_500_series-10        	    2686	    442439 ns/op	 1148513 B/op	   11523 allocs/op
Benchmark_Handler/PRW2_with_500_series
Benchmark_Handler/PRW2_with_500_series-10        	    1904	    619988 ns/op	 1423212 B/op	   16554 allocs/op
Benchmark_Handler/PRW1_with_1000_series
Benchmark_Handler/PRW1_with_1000_series-10       	    1360	    875548 ns/op	 2305732 B/op	   23027 allocs/op
Benchmark_Handler/PRW2_with_1000_series
Benchmark_Handler/PRW2_with_1000_series-10       	     952	   1241676 ns/op	 3047640 B/op	   33057 allocs/op
Benchmark_Handler/PRW1_with_2000_series
Benchmark_Handler/PRW1_with_2000_series-10       	     674	   1815843 ns/op	 4627024 B/op	   46032 allocs/op
Benchmark_Handler/PRW2_with_2000_series
Benchmark_Handler/PRW2_with_2000_series-10       	     478	   2528554 ns/op	 6302383 B/op	   66062 allocs/op
```

SungJin1212 commented:
I have noticed the Prometheus leave the log `we got 2xx status code from the Receiver yet statistics indicate some dat was not written; investigation needed ... failedSampleCount=0 failedHistogramCount=4 failedExemplarCount=0`.

SungJin1212 commented:
@yeya24 
Should we automatically enable `-blocks-storage.tsdb.enable-native-histograms`?
If we set the `protobuf_message` to `io.prometheus.write.v2.Request `, the Prometheus (Sender) send CT and NH by default. ([remote write config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write))

danielblando commented on pkg/distributor/distributor.go:
```diff
@@ -1202,6 +1214,13 @@ func (d *Distributor) send(ctx context.Context, ingester ring.InstanceDesc, time
 
+	if resp != nil {
+		// track write stats
+		ws.SetSamples(resp.Samples)
```
I was trying to think how to do this but it will be complex.

Lets say we have sample1, sample2 and sample3. And we have ing1,ing2,ing3,ing4 with replication 3

sample1 goes to ing1,ing2,ing3
sample2 goes to ing1,ing3,ing4
sample 3 goes to ing1,ing2,ing4

So send call will send this number of samples per ingester
ing1 - 3 samples
ing2 - 2 samples
ing3 - 2 samples
ing4 - 2 samples

If each of them returns 1 less than it should,
ing1 - 2
ing2 - 1
ing3 - 1
ing 4 - 1

It is very hard to say which samples were actually accepted. For example if sample 1 that went to ing1,ing2,ing3 was accepted in any of the ingester or if it was the one that failed in all 3 of them.

Makes sense? I might be wrong lol.

SungJin1212 replied:
It makes sense of your concern.. yeah it's hard to track how many samples are ingested accurately.
 
To simplify the problem, we could expose the count of validated samples, histograms, and exemplars as a header value from the distributor.

danielblando replied:
The validations on distributor are simple. I think the more business validations are on ingester. 

I was thinking and if we send the number with replication. So if we have a replication of 3 and distributor receives 1 samples, we return 3 on header.
It doesnt look that bad if we see the spec
```
X-Prometheus-Remote-Write-Samples-Written <count of all successfully written Samples>
```
3 is indeed the number of written samples

They also have 
```
More (optional) headers might come in the future, e.g. when more entities or fields are added and worth confirming.
``` 
in the spec
We could add an extra header with the replication number on it and that is a cortex header. Kind of extension of spec.

SungJin1212 replied:
Should we return 3 in this case?
When there are 3 ingesters and the Distributor sends a push request with 1 series, which has 1 sample. The two Ingesters accept it, but the other fails.

We could consider an approach where each Ingester adds the number of successful samples, and then the aggregated count is divided by the replication factor at the distributor side.
WDTY?

SungJin1212 replied:
I think it makes sense. Is there a way to reflect the stats from all 3 Ingesters? The response is returned when only `minSuccess` (2) Ingesters succeed.

danielblando replied:
True :( I didnt think about that. It is a weird behaviour. If now 1 ingester returns 4xx for a different sample we would have way more totalSamples returned because we have 3 responses instead of 2. Should we just not support this for now?

danielblando replied:
I am ok with that. It seems the best we can do to have consistent data for now.

SungJin1212 commented:
@danielblando 
Make sense of your concern. How about adding write stats at the Ingester side and dividing by # of replicationSet at the Distributor side?

For the above case, the header value would be (2+1+1+1)/4 = 1.
um.. It might not be clear.

yeya24 commented on docs/configuration/config-file-reference.md:
```diff
@@ -2894,6 +2894,11 @@ ha_tracker:
 
+# EXPERIMENTAL: If true, accept prometheus remote write v2 protocol push
+# request.
+# CLI flag: -distributor.remote-write2-enabled
```
Nit. Should the flag be named `remote-write-v2` instead of `remote-write2`?

danielblando reviewed (approved):
Amazing work
